### PR TITLE
feat: allow external cloud provides configration

### LINKED ISSUE #3312: Allow  --cloud-provider=external to be set
```yaml
cluster:
  externalCloudProvider:
    enabled: true
    manifests:
      - https://[…].yaml
```

If `enabled` is `true`, `manifests` list may be empty. If `enabled` is `false` (default), `manifests` list must be empty.

If enabled, `--cloud-provider=external` flag should be passed to kubelet, kube-apiserver, kube-controller-manager.

Manifests are merged into `extraManifests` on bootstrap.

### PR DESCRIPTION
**Pull request by AlekSi**
Closes #3312.

Requires changes/reorganization in cluster-api-templates. Draft: https://github.com/AlekSi/talos-cluster-api-templates/pull/1/files

- [x] you linked an issue (if applicable)
- [x] you included tests (if applicable)
- [x] you ran conformance (`make conformance`)
- [x] you formatted your code (`make fmt`)
- [x] you linted your code (`make lint`)
- [x] you generated documentation (`make docs`)
- [x] you ran unit-tests (`make unit-tests`)

### CODE REVIEW

File: internal/app/machined/pkg/controllers/config/k8s_control_plane.go [hunk 1]
```diff
@@ -145,10 +150,15 @@ func (ctrl *K8sControlPlaneController) manageAPIServerConfig(ctx context.Context
 }
 
 func (ctrl *K8sControlPlaneController) manageControllerManagerConfig(ctx context.Context, r controller.Runtime, logger *log.Logger, cfgProvider talosconfig.Provider) error {
+	var cloudProvider string
+	if cfgProvider.Cluster().ExternalCloudProvider().Enabled() {
+		cloudProvider = "external"
```
Comment (AlekSi) [inline internal/app/machined/pkg/controllers/config/k8s_control_plane.go:155]: Should be covered with test

File: internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go [hunk 2]
```diff
@@ -87,25 +87,10 @@ func (suite *K8sControlPlaneSuite) assertK8sControlPlanes(manifests []string) er
 	return nil
 }
 
-func (suite *K8sControlPlaneSuite) TestReconcileDefaults() {
+func (suite *K8sControlPlaneSuite) createMachine(cfg *config.MachineConfig) config.K8sControlPlaneAPIServerSpec {
```
Comment (AlekSi) [inline internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go:90]: Better method name?

File: internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go [hunk 3]
```diff
@@ -183,6 +171,52 @@ func (suite *K8sControlPlaneSuite) TestReconcileExtraVolumes() {
...
+					"https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/v1.20.0-alpha.0/manifests/rbac.yaml",
+					"https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/v1.20.0-alpha.0/manifests/aws-cloud-controller-manager-daemonset.yaml",
+				},
+			},
+		},
+	})
+
+	apiServerCfg := suite.createMachine(cfg)
+	suite.Assert().Equal("external", apiServerCfg.CloudProvider)
+
```
Comment (AlekSi) [inline internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go:199]: Also test controller manager config

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go [hunk 4]
```diff
@@ -207,6 +206,10 @@ func (ctrl *ControlPlaneStaticPodController) manageAPIServer(ctx context.Context
 		"--kubelet-preferred-address-types=InternalIP,ExternalIP,Hostname",
 	}
 
+	if cfg.CloudProvider != "" {
+		args = append(args, fmt.Sprintf("--cloud-provider=%s", cfg.CloudProvider))
```
Comment (AlekSi) [inline internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:210]: Should be covered by test

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go [hunk 5]
```diff
@@ -297,6 +299,10 @@ func (ctrl *ControlPlaneStaticPodController) manageControllerManager(ctx context
 		"--profiling=false",
 	}
 
+	if cfg.CloudProvider != "" {
+		args = append(args, fmt.Sprintf("--cloud-provider=%s", cfg.CloudProvider))
```
Comment (AlekSi) [inline internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:303]: Should be covered by test

File: internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go (hunk 2 above)
Comment (smira) [reply internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go:90]: `setupInputs`, as this is dependencies today, but it's going to be renamed to `Inputs` in #3405

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go [hunk 6]
```diff
@@ -1190,6 +1193,28 @@ type CNIConfig struct {
...
+	//   description: |
+	//     A list of urls that point to additional manifests for an external cloud provider.
+	//     These will get automatically deployed as part of the bootstrap.
+	//   examples:
+	//     - value: >
+	//        []string{
+	//         "https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/v1.20.0-alpha.0/manifests/rbac.yaml",
+	//         "https://raw.githubusercontent.com/kubernetes/cloud-provider-aws/v1.20.0-alpha.0/manifests/aws-cloud-controller-manager-daemonset.yaml",
+	//        }
+	Manifests []string `yaml:"manifests,omitempty"`
```
Comment (smira) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1215]: for consistency sake, should we name variable `ExtraManifests` and interface method `Manifests`?

**Review by smira — COMMENTED** [review]
lgtm, we need some tests via e2e-aws and small nits

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go (hunk 6 above)
Comment (AlekSi) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1215]: Should we rename the YAML config field too? How "extra" are those manifests? My understanding is that required more often than not, and those are not "extra". And then I would keep the struct field name the same as the config field name.

Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1215]: sorry, `ExternalCloudProviderManfiests` or whatever, as we usually prefix field names with struct name, while interface functions are short names

File: internal/app/machined/pkg/controllers/config/k8s_control_plane.go (hunk 1 above)
Comment (AlekSi) [reply internal/app/machined/pkg/controllers/config/k8s_control_plane.go:155]: Covered now

File: internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go (hunk 3 above)
Comment (AlekSi) [reply internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go:206]: Done

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go (hunk 4 above)
Comment (AlekSi) [reply internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:210]: Covered now

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go (hunk 5 above)
Comment (AlekSi) [reply internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:303]: Covered now

File: internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go (hunk 2 above)
Comment (AlekSi) [reply internal/app/machined/pkg/controllers/config/k8s_control_plane_test.go]: Renamed to `setupMachine`

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go (hunk 6 above)
Comment (AlekSi) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go]: Renamed field to `ExternalManifests`. Method name is `ManifestURLs` for consistency with existing `ExtraManifestURLs`.

File: hack/test/e2e-aws.sh [hunk 7]
```diff
@@ -78,8 +78,7 @@ function setup {
   ## TODO: update to talos-systems once merged
   ${CLUSTERCTL} config cluster ${NAME_PREFIX} \
     --kubeconfig /tmp/e2e/docker/kubeconfig \
-    --from https://github.com/talos-systems/cluster-api-templates/blob/main/aws/standard/standard.yaml > ${TMP}/cluster.yaml
-  
+    --from https://github.com/AlekSi/talos-cluster-api-templates/blob/external-cloud-provider/aws/standard-external/standard.yaml > ${TMP}/cluster.yaml
```
Comment (AlekSi) [inline hack/test/e2e-aws.sh:81]: TODO

File: hack/test/e2e-aws.sh [hunk 8]
```diff
@@ -78,8 +78,7 @@ function setup {
   ## TODO: update to talos-systems once merged
```
Comment (AlekSi) [inline hack/test/e2e-aws.sh:78]: I wonder what this comment means

Comment (smira) [reply hack/test/e2e-aws.sh:78]: I guess it's outdated as the templates lived under someone's else account

File: hack/test/e2e-aws.sh [hunk 9]
```diff
@@ -78,8 +78,7 @@ function setup {
...
   ${CLUSTERCTL} config cluster ${NAME_PREFIX} \
     --kubeconfig /tmp/e2e/docker/kubeconfig \
-    --from https://github.com/talos-systems/cluster-api-templates/blob/main/aws/standard/standard.yaml > ${TMP}/cluster.yaml
```
Comment (smira) [inline hack/test/e2e-aws.sh:81]: can we revert this change, merge this PR

then refactor templates to use versions via folders, and flip the template here to 0.2 version with cloud-provider external from the config as a separate PR?

**Review by smira — APPROVED** [review]

File: hack/test/e2e-aws.sh (hunk 9 above)
Comment (AlekSi) [reply hack/test/e2e-aws.sh:81]: Done

**Review by smira — APPROVED** [review]
